Dockerfile

```diff
@@ -1,39 +1,143 @@
-RUN chown deploy:deploy $(pwd)
+# Copy files required for installation of application dependencies
+COPY package.json yarn.lock ./
+COPY patches ./patches
 
-# Switch to less-privileged user, do this early to ensure files are created
-# with proper ownership.
-USER deploy
+# ---------------------------------------------------------
+# Yarn dependencies
+# ---------------------------------------------------------
+FROM yarn-base as yarn-deps
 
-# Copy files required for installation of application dependencies
-COPY --chown=deploy:deploy package.json yarn.lock ./
-COPY --chown=deploy:deploy patches ./patches
+RUN yarn install --production --frozen-lockfile --quiet \
+  && mv node_modules /opt/node_modules.prod \
+  && yarn install --frozen-lockfile --quiet \
+  && yarn cache clean --force
 
-# Install application dependencies
-RUN yarn install --frozen-lockfile --quiet && \
-  yarn cache clean --force
+# ---------------------------------------------------------
+# Builder with source code
+# ---------------------------------------------------------
+FROM yarn-deps as builder-src
 
 # Copy application code
-COPY --chown=deploy:deploy . ./
+COPY __mocks__ ./__mocks__
+COPY cypress ./cypress
+COPY data ./data
+COPY patches ./patches
+COPY src ./src
+COPY webpack ./webpack
+COPY .env.oss \
+  .env.test \
+  .eslintrc.js \
+  .nvmrc \
+  .prettierignore \
+  apollo.config.js \
+  babel.config.js \
+  coffeelint.json \
+  cypress.json \
+  dangerfile.ts \
+  jest.config.js \
+  package.json \
+  relay.config.js \
+  renovate.json \
+  test.config.js \
+  test.mocha.js \
+  tsconfig.json \
+  yarn.lock \
+  ./
+
+# ---------------------------------------------------------
+# Compile assets
+# ---------------------------------------------------------
+FROM builder-src as builder-assets
+
+# Build application
+RUN yarn assets
+
+# ---------------------------------------------------------
+# Compile server
+# ---------------------------------------------------------
+FROM builder-src as builder-server
 
 # Build application
-RUN yarn assets && \
-    yarn build:server
+RUN yarn build:server
+
+# ---------------------------------------------------------
+# All development assets
+# ---------------------------------------------------------
+FROM builder-src as builder
+
+# Scripts
+COPY ./scripts ./scripts
+
+# Client assets
+COPY --from=builder-assets /app/manifest.json .
+COPY --from=builder-assets /app/public ./public
+COPY --from=builder-assets /app/src ./src
+
+# Server assets
+COPY --from=builder-server /app/server.dist.js .
+COPY --from=builder-server /app/server.dist.js.map .
+
+# ---------------------------------------------------------
+# Release image
+# ---------------------------------------------------------
+#
+# Release stage. This stage creates the final docker iamge that will be
+# released. It contains only production dependencies and artifacts.
+#
+# TODO: Make the COPY steps more explicit, the iamge still copies in the all raw
```

Comment:
trivial, `iamge` typo

Reply:
Thanks for finding this. I was able to remove the TODO completely!